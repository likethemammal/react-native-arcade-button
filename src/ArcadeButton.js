import React, { Component } from 'react'
import {
    StyleSheet,
    Text,
    View,
    Animated,
    Touchable,
} from '@likethemammal/react-primitives'

import colors from './constants/colors'

const reflectionColor = 'rgba(255,255,255,0.35)'
const shadowColor = 'rgba(0,0,0,0.1)'
const shadowColorDark = 'rgba(0,0,0,0.15)'
const bottomShadowColor = colors.silver

export default class ArcadeButton extends Component {

    static defaultProps = {
        color: 'hsla(120, 52%, 48%, 1)',
    }

    state = {
        down: false,
        amountDown: new Animated.Value(0),
    }

    animationTimer = false

    onDown = () => {

        if (this.state.down) {
            return
        }

        const { onDown, disabled } = this.props

        this.setState({
            down: true
        })

        if (disabled) {
            return
        }

        this.animationTimer = Animated.timing(          // Uses easing functions
            this.state.amountDown,    // The value to drive
            {
                toValue: 1,
                delay: 0,
                duration: 42
            }            // Configuration
        );

        this.animationTimer.start()

        if (onDown) {
            onDown()
        }
    }

    onUp = () => {
        const { onUp, disabled, onDisabled, } = this.props

        this.setState({
            down: false
        })

        if (disabled) {
            if (onDisabled) onDisabled()
            return
        }

        //react native doesnt accect the value being reset to a new value of zero
        this.state.amountDown.setValue(0)

        if (onUp) {
            onUp()
        }
    }

    componentDidUpdate(prevProps, prevState) {
        if (!prevProps.disabled && this.props.disabled) {
            this.state.amountDown.setValue(1)
        }

        if (prevProps.disabled && !this.props.disabled) {
            this.state.amountDown.setValue(0)
        }
    }

    render() {

        const { amountDown } = this.state
        const { color, text } = this.props

        const topDepressed = {
            transform: [{
                translateY: amountDown.interpolate({
                    inputRange: [0, 1],
                    outputRange: [0, 2]
                })
            }],
            zIndex: 1,
            opacity: amountDown.interpolate({
                inputRange: [0, 1],
                outputRange: [1, 0.8]
            })
        }

        const disabled = this.props.disabled && disabledStyles

        const topFlat = {
            backgroundColor: color,
            borderColor: color,
        }
        const bottomFlat = {
            backgroundColor: color,
        }
        const bottomRim = {
            backgroundColor: color,
        }

        return (
            <View style={styles.container}>
                <View style={styles.bottom}>
                    <View style={[styles.bottomShadowUnder, disabled && disabled.bottomShadowUnder]} />
                    <View style={[styles.bottomRim, bottomRim, disabled && disabled.bottomFlat]} />
                    <View style={[styles.bottomShadow, disabled && disabled.bottomShadow]} />
                    <View style={[styles.bottomFlat, bottomFlat, disabled && disabled.bottomFlat]} />
                </View>

                <View style={styles.top}>

                    <View style={[styles.topShadow, disabled && disabled.topShadow]} />

                    <Animated.View style={topDepressed}>

                        <View style={[styles.topReflection, disabled && disabled.topReflection]} />

                        <Touchable onPressIn={this.onDown} onPressOut={this.onUp}>
                            <View style={[styles.topFlat, topFlat, disabled && disabled.topFlat]}>
                                <Text style={[styles.text, disabled && disabled.text]}>
                                    {text}
                                </Text>
                            </View>
                        </Touchable>

                    </Animated.View>

                </View>

            </View>
        );
    }
}

const disabledStyles = StyleSheet.create({
    topFlat: {
        backgroundColor: colors.silver_dark,
        borderColor: colors.silver_dark,
    },
    topReflection: {
        backgroundColor: colors.silver,
    },
    bottomFlat: {
        borderColor: colors.silver,
        backgroundColor: colors.silver_dark,
    },
    topShadow: {
        borderColor: colors.gray,
        backgroundColor: colors.dusty_gray,
    },
    bottomShadow: {
        backgroundColor: colors.dusty_gray,
    },
    bottomShadowUnder: {
        backgroundColor: bottomShadowColor,
    },
    text: {
        textShadowColor: 'rgba(0,0,0,0.0)',
        opacity: 0.5,
    }
})

const styles = StyleSheet.create({
    container: {
        alignItems: 'center',
        position: 'relative',
        top: 3,
        width: 220,
        height: 220,
    },
    top: {
        position: 'relative',
        top: -3,
    },
    topFlat: {
        alignItems: 'center',
        justifyContent: 'center',
        width: 150,
        height: 150,
        marginTop: 3,
        borderRadius: 75,
        borderWidth: 1,
        borderStyle: 'solid',
    },
    text: {
        color: 'white',
        fontWeight: 'bold',
        fontSize: 32,
        textShadowOffset: {
            width: 0,
            height: 1
        },
        textShadowColor: 'rgba(0,0,0,0.2)',
        textShadowRadius: 5,
        textTransform: 'uppercase'
    },
    topShadow: {
        position: 'absolute',
        width: 150,
        height: 150,
        marginTop: 7,
        borderRadius: 75,
        borderColor: shadowColorDark,
        borderWidth: 1,
        borderStyle: 'solid',
        backgroundColor: shadowColor,
    },
    topReflection: {
        position: 'absolute',
        width: 150,
        height: 150,
        marginTop: 1,
        borderRadius: 75,
        backgroundColor: reflectionColor,
    },
    topDepressed: {
    },
    bottom: {
        position: 'relative',
        marginBottom: -175,
    },
    bottomFlat: {
        width: 200,
        height: 200,
        marginTop: 5,
        borderRadius: 100,
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: reflectionColor,
    },
    bottomRim: {
        position: 'absolute',
        top: 8,
        width: 200,
        height: 200,
        borderRadius: 100,
    },
    bottomShadow: {
        position: 'absolute',
        top: 8,
        width: 200,
        height: 200,
        borderRadius: 100,
        backgroundColor: shadowColor,
    },
    bottomShadowUnder: {
        position: 'absolute',
        top: -3,
        left: -10,
        width: 220,
        height: 220,
        borderRadius: 110,
        backgroundColor: bottomShadowColor,
        opacity: 0.1,
    }
})
