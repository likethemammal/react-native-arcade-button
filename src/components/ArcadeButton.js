import React, { Component } from 'react'
import {
    StyleSheet,
    Text,
    View,
    Animated,
    Touchable,
} from '@likethemammal/react-primitives'

import _ from 'underscore'

import onecolor from 'onecolor'

import colors from '../constants/colors'

const reflectionColor = colors.malachite
const flatColor = colors.jade
const shadowColor = colors.green_haze
const shadowColorDark = colors.green_haze_dark
const bottomShadowColor = colors.silver

function hslToString({ _hue, _saturation, _lightness}) {

    return `hsl(${_hue*360}, ${_saturation*100}%, ${_lightness*100}%)`
}

const originalColors = {
    flatColor,
    reflectionColor,
    shadowColor,
    shadowColorDark,
}

const originalColorsHSLs = _.mapObject(
    originalColors,
    (val, key) => onecolor(val).hsl()
)
const newColor = colors.jade

const newColorHSL = onecolor(newColor).hsl()

const differenceColorsHSLs = _.mapObject(
    originalColorsHSLs,
    ({
        _hue,
        _saturation,
        _lightness,
    }, key) => {
        return {
            _hue: originalColorsHSLs.flatColor._hue - _hue,
            _saturation: originalColorsHSLs.flatColor._saturation - _saturation,
            _lightness: originalColorsHSLs.flatColor._lightness - _lightness,
        }
    }
)

const newColorsHSLs = _.mapObject(
    differenceColorsHSLs,
    ({
        _hue,
        _saturation,
        _lightness,
    }, key) => {
        return {
            _hue: newColorHSL._hue - _hue,
            _saturation: newColorHSL._saturation - _saturation,
            _lightness: newColorHSL._lightness - _lightness,
        }
    }
)

const newColors = _.mapObject(
    newColorsHSLs,
    hslToString
)


export default class ArcadeButton extends Component {
    state = {
        down: false,
        amountDown: new Animated.Value(0),
    }

    animationTimer = false

    onDown = () => {

        if (this.state.down) {
            return
        }

        const { onDown, disabled } = this.props

        this.setState({
            down: true
        })

        if (disabled) {
            return
        }

        this.animationTimer = Animated.timing(          // Uses easing functions
            this.state.amountDown,    // The value to drive
            {
                toValue: 1,
                delay: 0,
                duration: 42
            }            // Configuration
        );

        this.animationTimer.start()

        if (onDown) {
            onDown()
        }
    }

    onUp = () => {
        const { onUp, disabled, onDisabled, } = this.props

        this.setState({
            down: false
        })

        if (disabled) {
            onDisabled()
            return
        }

        //react native doesnt accect the value being reset to a new value of zero
        this.state.amountDown.setValue(0)

        if (onUp) {
            onUp()
        }
    }

    componentDidUpdate(prevProps, prevState) {
        if (!prevProps.disabled && this.props.disabled) {
            this.state.amountDown.setValue(1)
        }

        if (prevProps.disabled && !this.props.disabled) {
            this.state.amountDown.setValue(0)
        }
    }

    render() {
        const topDepressed = {
            transform: [{
                translateY: this.state.amountDown.interpolate({
                    inputRange: [0, 1],
                    outputRange: [0, 2]
                })
            }],
            zIndex: 1,
            opacity: this.state.amountDown.interpolate({
                inputRange: [0, 1],
                outputRange: [1, 0.8]
            })
        }

        const disabled = this.props.disabled && {
                topFlat: {
                    backgroundColor: colors.silver_dark,
                    borderColor: colors.silver_dark,
                },
                topReflection: {
                    backgroundColor: colors.silver,
                },
                bottomFlat: {
                    borderColor: colors.silver,
                    backgroundColor: colors.silver_dark,
                },
                topShadow: {
                    borderColor: colors.gray,
                    backgroundColor: colors.dusty_gray,
                },
                bottomShadow: {
                    backgroundColor: colors.dusty_gray,
                },
                bottomShadowUnder: {
                    backgroundColor: bottomShadowColor,
                },
                text: {
                    textShadowColor: 'rgba(0,0,0,0.0)',
                    opacity: 0.5,
                }
            }

        return (
            <View style={styles.container}>
                <View style={styles.bottom}>
                    <View style={[styles.bottomShadowUnder, disabled && disabled.bottomShadowUnder]} />
                    <View style={[styles.bottomShadow, disabled && disabled.bottomShadow]} />
                    <View style={[styles.bottomFlat, disabled && disabled.bottomFlat]} />
                </View>

                <View style={styles.top}>

                    <View style={[styles.topShadow, disabled && disabled.topShadow]} />

                    <Animated.View style={topDepressed}>

                        <View style={[styles.topReflection, disabled && disabled.topReflection]} />

                        <Touchable onPressIn={this.onDown} onPressOut={this.onUp}>
                            <View style={[styles.topFlat, disabled && disabled.topFlat]}>
                                <Text style={[styles.text, disabled && disabled.text]}>Start</Text>
                            </View>
                        </Touchable>

                    </Animated.View>

                </View>

            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        position: 'relative',
        alignItems: 'center',
        marginBottom: 30,
        marginTop: 30,
    },
    top: {
        position: 'relative',
        top: -3,
    },
    topFlat: {
        alignItems: 'center',
        justifyContent: 'center',
        width: 150,
        height: 150,
        marginTop: 3,
        borderRadius: 75,
        backgroundColor: newColors.flatColor,
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: newColors.flatColor,
    },
    text: {
        color: 'white',
        fontWeight: 'bold',
        fontSize: 30,
        textShadowOffset: {
            width: 0,
            height: 1
        },
        textShadowColor: 'rgba(0,0,0,0.2)',
        textShadowRadius: 5,
    },
    topShadow: {
        position: 'absolute',
        width: 150,
        height: 150,
        marginTop: 7,
        borderRadius: 75,
        borderColor: newColors.shadowColorDark,
        borderWidth: 1,
        borderStyle: 'solid',
        backgroundColor: newColors.shadowColor,
    },
    topReflection: {
        position: 'absolute',
        width: 150,
        height: 150,
        marginTop: 1,
        borderRadius: 75,
        backgroundColor: newColors.reflectionColor,
    },
    topDepressed: {
    },
    bottom: {
        position: 'relative',
        marginBottom: -175,
    },
    bottomFlat: {
        width: 200,
        height: 200,
        marginTop: 5,
        borderRadius: 100,
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: newColors.reflectionColor,
        backgroundColor: newColors.flatColor,
    },
    bottomShadow: {
        position: 'absolute',
        top: 8,
        width: 200,
        height: 200,
        borderRadius: 100,
        backgroundColor: newColors.shadowColor,
    },
    bottomShadowUnder: {
        position: 'absolute',
        top: -3,
        left: -10,
        width: 220,
        height: 220,
        borderRadius: 110,
        backgroundColor: bottomShadowColor,
        opacity: 0.15,
    }
})
