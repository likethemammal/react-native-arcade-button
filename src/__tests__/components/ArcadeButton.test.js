import React from 'react'
import { mount, shallow, } from 'enzyme'

import general_components from '@likethemammal/general-components'

const shared = general_components.__tests__._shared.tests

import ArcadeButton from '../../ArcadeButton'
import { Touchable } from 'react-primitives'

const mockOut_func = jest.fn()
const mockIn_func = jest.fn()
const multiple_mockIn_func = jest.fn()

const component = shallow(
    <ArcadeButton
        onPressIn={mockIn_func}
        onPressOut={mockOut_func}
    />
)

const multiple_component = mount(
    <ArcadeButton
        onPressIn={multiple_mockIn_func}
    />
)

const empty_component = shallow(
    <ArcadeButton />
)

const disabled_mockDisabled_func = jest.fn()
const disabled_mockOut_func = jest.fn()
const disabled_mockIn_func = jest.fn()

const disabled_component = mount(
    <ArcadeButton
        onDisabledOut={disabled_mockDisabled_func}
        onPressOut={disabled_mockOut_func}
        onPressIn={disabled_mockIn_func}
        disabled={true}
    />
)


describe('ArcadeButton', () => {

    shared.SNAPSHOTS_SHOULD_MATCH([
        component,
        disabled_component,
        empty_component,
        multiple_component,
    ])

    test(`if not disabled, mock funcs should  be called`, () => {
        component.find(Touchable).props().onPressIn()
        component.find(Touchable).props().onPressOut()

        expect(
            mockIn_func.mock.calls.length
        ).toEqual(1)

        expect(
            mockOut_func.mock.calls.length
        ).toEqual(1)
    })

    test(`if not disabled, calling onPressIn multiple times should only call the mock once`, () => {

        multiple_component.find(Touchable).props().onPressIn()
        multiple_component.find(Touchable).props().onPressIn()
        multiple_component.find(Touchable).props().onPressIn()

        expect(
            multiple_mockIn_func.mock.calls.length
        ).toEqual(1)
    })

    test(`if disabled, onPressIn should not be called`, () => {
        disabled_component.find(Touchable).props().onPressIn()

        expect(
            disabled_mockIn_func.mock.calls.length
        ).toEqual(0)
    })

    test(`if disabled, onPressOut should not be called`, () => {
        disabled_component.find(Touchable).props().onPressOut()

        expect(
            disabled_mockOut_func.mock.calls.length
        ).toEqual(0)
    })

    test(`if disabled, onPressOut should call onDisabledOut mock function`, () => {
        disabled_component.find(Touchable).props().onPressOut()

        expect(
            disabled_mockDisabled_func.mock.calls.length
        ).toEqual(2)

        //twice, because of the earlier onPressOut
    })

})