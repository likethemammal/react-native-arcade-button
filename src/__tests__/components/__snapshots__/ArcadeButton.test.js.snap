// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxWithLabel Snapshot 0 should match 1`] = `undefined`;
